### DISCUSSION TITLE
CODEOWNER Required Reviews

### DISCUSSION THREAD

**Discussion by benknoble** [community/community#22522] [New to GitHub]
<p>I’ve got a repository with 3 total collaborators, and the master/dev branches are set up to be:</p>
<ul>
<li>protected</li>
<li>require PR reviews</li>
<li>require review from Code Owners</li>
</ul>
<p>Now, as I understood things, my open PR towards the dev branch should request a review from all Code Owners, but a single approved review should be enough to review.</p>
<p>My PR automatically requested review from the other two collaborators, since both are Code Owners. So far, so good. One of them requested changes, which we discussed, made, and then approved. So the current status of his review is “approved”–yet this is not enough to merge the PR?</p>
<p>The status box says</p>
<p>Code owner review required<br>
Waiting on code owner review from xxxxxxxxxxxx.</p>
<p>Required statuses must pass before mergingAll required <a href="https://developer.github.com/v3/repos/statuses/" rel="nofollow noopener">status checks</a> on this pull request must run successfully to enable automatic merging.</p>
<p>There are no status checks enabled (beyond required reviews), and I have <em>at least one approved review</em>, so I fail to see why the merge button didn’t turn that shiny Green (I have full privilege on the repo, so presumably I can merge when the review is approved).</p>
<p>Is there any way to make sure at least one approved review is enough to merge the PR, or is this one of those “since it’s your PR you can’t merge it” things?</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>It’s hard to say why you’re experiencing this behavior without some more information. Is the repository in question public? Can you share a link to a specific PR that is behaving in a way that you believe is not expected?</p>
<p>Assuming that the repository in question is private, I can guess at one reason why more than one reviewer might be required. Let’s say you have a CODEOWNERS like the following:</p>
<pre><code>* @foo @bar @baz
</code></pre>
<p>In this case, any PR would require one (and only one) of those three people to approve it before it could be merged. But let’s say you have a more complex CODEOWNERS file:</p>
<pre><code>*.js @foo @bar
docs/ @baz
</code></pre>
<p>And you submit a PR that includes both a JavaScript change as well as a change to some associated documentation. That PR would require one of either foo or bar to approve it for the JavaScript change and a separate approval from baz for the documentation.</p>
<p>Maybe you could share just your CODEOWNERS file?</p>

**Comment by benknoble** [comment]
<p>I figured it out; it was similar to what you describe below, where my changes were targeting two different patterns, requiring a review from two separate people to approve each change.</p>
<p>Thanks for your help!</p>

**Comment by jpe442** [comment]
<p>Elsewhere you have said you can specify code owners for specific branches. Can you please give an example?</p>

**Comment by lee-dohm** [comment]
<p>There is an example in <a href="https://help.github.com/articles/about-codeowners/#codeowners-file-location" rel="nofollow noopener">the CODEOWNERS help documentation</a>. Let us know if you have any questions about it.</p>

**Comment by JasonCorns-Avalara** [comment]
<p>hey <a class="mention" href="/lee-dohm">@lee-dohm</a></p>
<p>I don’t see any code example.  I found a blurb about branches, but nothing telling us <em>how</em></p>
<blockquote>
<p>Each CODEOWNERS file assigns the code owners for a single branch in the repository. Thus, you can assign different code owners for different branches, such as <span class="mention">@octo-org</span>/codeowners-team for a code base on the master branch and <span class="mention">@octocat</span> for a GitHub Pages site on the gh-pages branch.</p>
</blockquote>
<p>on the docs page; what does an actual branch reference look like?  would we have a new CODEOWNERS file per branch, or would we have the same CODEOWNERS file with all the branches and reviewers listed?</p>

**Comment by lee-dohm** [comment]
<p>It’s probably better explained by saying that the one CODEOWNERS file has different contents per branch. So if there is one rule in the file on the <code>master</code> branch but a different rule in the file on the <code>foo</code> branch, then the first rule applies when things are being <em>merged into</em> <code>master</code> and the second rule applies when things are being <em>merged into</em> <code>foo</code>. In both cases, the format of the CODEOWNERS file is the same and there is no mention of a branch reference within the contents of the CODEOWNERS file itself.</p>

**Comment by sethduda** [comment]
<pre><code>* @foo @bar @baz
</code></pre>
<p>“In this case, any PR would require one (and only one) of those three people to approve it before it could be merged”</p>
<p>Is there any way to require <strong>all</strong> of the listed people to approve?</p>

**Comment by lee-dohm** [comment]
<p>No, there currently isn’t a way to do that built-in to the CODEOWNERS feature. If that was the one and only rule, you could do a hacky solution where you required a PR to have three approvals to be merged but if the CODEOWNERS contents ever changed, you would have to remember to manually update the number of required reviews. So it isn’t something I would recommend as a solution to the question you’re asking.</p>

**Comment by robdodson** [comment]
<p>Is there a way to make it so the universal code owner can approve PRs to subdirectories that also have code owners. For example:</p>
<pre><code class="lang-auto">\*&amp;nbsp;@foo  
  
/docs&amp;nbsp;@bar&amp;nbsp;@baz  
</code></pre>
<p>Let’s say someone does a PR to the <code>docs/</code> directory to fix a missing comma. While I like that bar and baz get requested for reviews on the <code>docs/</code> directory, it seems like if foo adds themselves as a reviewer and approves the PR then that should satisfy the “1 owner approves” rule.</p>
<p>I realize I can work around this by adding foo as an owner to <code>docs/</code> but in a project with many owners it seems a bit inconvenient to not rely on a pattern when an owner can approve PRs to their directory and all subdirectories. Additional owners on subdirectories don’t override the ownership of the person higher up in the hierarchy—they just get added in.</p>

**Reply by tleerai** [reply]
Is there any plan to add this functionality?

The idea would be that a specific rule could require an approval from two groups or two people:

`/.github @devops && @devteam`

**Comment by lee-dohm** [comment]
<p>I can understand the rationale behind what you’re describing. However, that isn’t how the feature is currently implemented. But even if the rules were additive as you describe, it wouldn’t only notify <code>@bar</code> and <code>@baz</code>, <code>@foo</code> would get notified as well. The workflow you describe appears to indicate that <code>@foo</code> wouldn’t get notified but would still count as an owner if they review the PR.</p>
<p>For workflows customized to your particular development team, I would recommend investigating <a href="https://probot.github.io" rel="nofollow noopener">Probot</a> or <a href="https://github.com/features/actions" rel="nofollow noopener">GitHub Actions</a> to see if either of them can help you build a solution to fit your team’s needs.</p>

**Comment by mderazon** [comment]
<p>So how do you avoid the PR merge from one branch to another overwrite the CODEOWNERS file with different content?</p>

**Comment by duartec003** [comment]
<p>update the .gitignore file in all branches to include CODEOWNERS</p>
<p>and from there you can specify different codeowners that won’t overwrite on merge</p>

**Comment by logankilpatrick** [comment]
<p>Bumping this: <a class="mention" href="/lee-dohm">@lee-dohm</a> </p>

**Comment by lee-dohm** [comment]
<p>I’m not sure why you’re bumping this <a class="mention" href="/logankilpatrick">@logankilpatrick</a>.</p>

**Comment by logankilpatrick** [comment]
<p>Hey! Sorry, it does not look like my bump was in the right place, but the final question on the thread was whether and how you could set it up such that you require multiple code owners to approve the PR.  Thanks! </p>

**Comment by lee-dohm** [comment]
<p><a href="https://github.community/t/codeowner-required-reviews/356/9">My post on 28 November, 2019</a> is in response to the question about how to require multiple owners to review a PR before it could be merged.</p>
<p>As I stated in <a href="https://github.community/t/codeowner-required-reviews/356/11">my post above</a>, if the current CODEOWNERS implementation isn’t what you’re looking for, then I’d suggest investigating <a href="https://github.com/probot/probot" rel="nofollow noopener">Probot</a> or <a href="https://github.com/features/actions" rel="nofollow noopener">GitHub Actions</a> to see if they’ll allow you to model the workflow you do want.</p>
<p>I hope that helps!</p>

**Comment by samajammin** [comment]
<p>A related question <a class="mention" href="/lee-dohm">@lee-dohm</a>: what happens if I require more PR approvals than there are code owners?</p>
<p>Our use case: we want to require code owner review on a specific file. e.g CODEOWNERS file:</p>
<pre><code class="lang-auto">some-important-file.js @foo @bar @baz
</code></pre>
<p>From my understanding, these 3 users would be codeowners for that file but if a PR doesn’t touch that file, there are 0 code owners.</p>
<p>In this situation, is Github’s logic smart enough to not require approvals on a PR that only touches files <em>without</em> any owners? Or would every PR still require 3 approvals?<br>
Is there any way to achieve my desired outcome: to only require PR approvals on PRs that touch a specific file?</p>
<p>Thanks in advance!</p>

**Comment by lee-dohm** [comment]
<p>Great question <a class="mention" href="/samajammin">@samajammin</a> :+1:</p>
<p>Requiring a specific number of reviews means that every PR must have that number of reviews before it can be merged. The logic will not forego requiring a PR review on PRs that have no files covered by <code>CODEOWNERS</code>.</p>
<aside class="quote no-group" data-username="samajammin" data-post="18" data-topic="356">
<div class="title">
<div class="quote-controls"></div>
 samajammin:</div>
<blockquote>
<p>Is there any way to achieve my desired outcome: to only require PR approvals on PRs that touch a specific file?</p>
</blockquote>
</aside>
<p>As I mentioned above, the general solution to having any sort of custom check logic that allows or prevents merging PRs is to use GitHub Actions or some similar system like Probot.</p>

**Comment by tisonkun** [comment]
<p>Hi <a class="mention" href="/lee-dohm">@lee-dohm</a> I’d like to ask the situation when PR matches multiple distinguished rules. Said that we have a <code>CODEOWNERS</code> files like:</p>
<pre><code class="lang-auto">*.js @foo @bar
docs/ @bar @baz
</code></pre>
<p>If someone make a PR touches both JS files and docs folder, is GitHub choose only <code>@bar</code> as the reviewer, or it must require two different reviewers, or some other behavior?</p>

**Reply by Alexander-Shukaev** [reply]
bump

**Reply by endersonmenezes** [reply]
I had a similar use case at my company. With a GitHub App that could read the company's teams, I started creating some solutions that gave it "extra power" and configured it as a Required Workflow. It's not ideal, but it helped us implement "powers" in the CODEOWNERs file.

https://github.com/endersonmenezes/codeowners-superpowers
